### PR TITLE
Implement AB testing for Matej recommendation models

### PR DESCRIPTION
**Pull request by foglcz**
|  |  |
| --- | --- |
| **Type** | Feature |
| **Fixes issues** | RAD-1066 |
| **Documentation** | updated |
| **BC Break** | no |
| **Tests updated** | yes |

Contains implementation for A/B testing of various recommendation models. These models have to be implemented/deployed in Matej first, before they're available for any given client. This is by design, as new model requires full recalculation, and therefore availability is individual depending on the size of clients data.

By default, only `default` model is available for every client (which means *clients* default, not *matej* default.)

Requires **Matej release ~v7.37~ v7.36**
Currently deployed to Staging instance, which contains models `default` and `model_b`.

There is no API endpoint that'd list available models. No validation beyond basic identifier is required.

### CODE REVIEW

File: src/Model/Command/Sorting.php [hunk 1]
```diff
@@ -14,21 +14,40 @@ class Sorting extends AbstractCommand implements UserAwareInterface
     private $userId;
     /** @var string[] */
     private $itemIds = [];
+    /** @var string|null */
+    private $modelName = null;
 
-    private function __construct(string $userId, array $itemIds)
+    private function __construct(string $userId, array $itemIds, ?string $modelName = null)
```
Comment (jirinovak) [inline src/Model/Command/Sorting.php:20]: Maybe we can use `default` as default value and request `string`. As a benefit we can get rid of all null checks in other methods.

File: tests/integration/RequestBuilder/EventsRequestBuilderTest.php [hunk 2]
```diff
@@ -6,13 +6,37 @@
...
+use Lmc\Matej\Model\Command\ItemPropertySetup;
 use Lmc\Matej\Model\Command\UserMerge;
+use Lmc\Matej\RequestBuilder\ItemPropertiesSetupRequestBuilder;
 
 /**
  * @covers \Lmc\Matej\RequestBuilder\EventsRequestBuilder
  */
 class EventsRequestBuilderTest extends IntegrationTestCase
 {
+    private function buildAndSendPropertySetupRequest(ItemPropertiesSetupRequestBuilder $builder): void
```
Comment (jirinovak) [inline tests/integration/RequestBuilder/EventsRequestBuilderTest.php:18]: Commit 95c55be86ff6a93469e0251226ea8393f2486bd0 seems to be unrelated with this PR and should be separate PR. Or am I wrong?

File: tests/integration/RequestBuilder/EventsRequestBuilderTest.php [hunk 3]
```diff
@@ -6,13 +6,37 @@
...
+    {
+        $builder->addProperty(ItemPropertySetup::string('test_property_a'))
+            ->addProperty(ItemPropertySetup::string('test_property_b'))
+            ->addProperty(ItemPropertySetup::string('test_property_c'))
+            ->send();
+    }
+
+    protected function setup(): void
```
Comment (jirinovak) [inline tests/integration/RequestBuilder/EventsRequestBuilderTest.php:26]: `setup` and `tearDown` methods should be first.

And helper method `buildAndSendPropertySetupRequest` should be at the end of test class.

**Review by jirinovak — CHANGES_REQUESTED** [review]

**Comment by kubasimon** [comment]
What is the motivation to make client responsible for selecting right 'model' (a/b testing) and not to select right model (or a/b test models) on server? (especially when new model "by design requires full recalculation")

**Comment by VincTheSecond** [comment]
@tenerd Currently, Matej is not able to measure business metrics (click rates, conversions) exactly as it requires clients to send us something like RPS codes for all interactions. In 2018.Q2 we will introduce a new strategy how to do it. After that, we will able to provide AB testing exclusively in Matej, without any client participation. For now, this is a temporary solution for clients who want to test different models even in 2018.Q2 :-)

File: README.md [hunk 4]
```diff
@@ -251,6 +251,51 @@ $response = $matej->request()
...
+$sorting->setModelName('custom_name_of_model_b');
+
+$response = $matej->request()
+    ->sorting($sorting)
+    ->send();
+```
+
+`custom_name_of_model_b` will be provided to you by LMC.
+
+If you use `NULL`, or `'default'`, the default model for your instance will be used.
```
Comment (OndraM) [inline README.md:294]: `null` in lowercase

File: README.md [hunk 5]
```diff
@@ -251,6 +251,51 @@ $response = $matej->request()
...
+
+A/B testing is available in `/recommendation` and `/sorting` endpoints for individual requests,
```
Comment (OndraM) [inline README.md:296]: I'd not refer to endpoints in the "URL" form, rather this could by hyperlink to the respective query builders:
`... is available in [Recommendation](#recommendations-for-single-user), ..."`

File: README.md [hunk 6]
```diff
@@ -251,6 +251,51 @@ $response = $matej->request()
...
+as well as in `/campaign` endpoint for batch requests.
```
Comment (OndraM) [inline README.md:297]: "... as well as in bath request created via `campaign()` builder."

File: src/Model/Command/Sorting.php (hunk 1 above)
Comment (OndraM) [reply src/Model/Command/Sorting.php:20]: The question is whether Matej behavior will be the same for `null` and `"default"`, @foglcz ?

File: src/Model/Command/Sorting.php [hunk 7]
```diff
@@ -14,21 +14,40 @@ class Sorting extends AbstractCommand implements UserAwareInterface
...
+        $this->setModelName($modelName);
     }
 
     /**
      * Sort given item ids for user-based recommendations.
      *
      * @return static
      */
-    public static function create(string $userId, array $itemIds): self
+    public static function create(string $userId, array $itemIds, string $modelName = null): self
```
Comment (OndraM) [inline src/Model/Command/Sorting.php:32]: What about adding second static builder (with modelName) and keeping this one as is? `createWithModel()` etc.

Because Vincent wrote this is temporary solution, to be removed in the future once implemented inside Matej. And it is easier to remove (and find in codebase) on particular method call than method call with third parameter.

File: src/Model/Command/UserRecommendation.php [hunk 8]
```diff
@@ -51,16 +54,18 @@ private function __construct(string $userId, int $count, string $scenario, float
      * Set from 0.0 for no rotation (same items will be recommended) up to 1.0 (same items should not be recommended).
      * @param int $rotationTime Specify for how long will the item's rotationRate be taken in account and so the item
      * is penalized for recommendations.
+     * @param string|null $modelName Specify which model you want to use
```
Comment (OndraM) [inline src/Model/Command/UserRecommendation.php:57]: Dtto, what do you think about introducing second constructor?

File: tests/integration/RequestBuilder/EventsRequestBuilderTest.php (hunk 2 above)
Comment (OndraM) [reply tests/integration/RequestBuilder/EventsRequestBuilderTest.php:18]: Also, why is the test failing 🤔?

File: tests/integration/RequestBuilder/RecommendationRequestBuilderTest.php [hunk 9]
```diff
@@ -20,7 +21,7 @@ public function shouldExecuteRecommendationRequestOnly(): void
     {
         $response = $this->createMatejInstance()
             ->request()
-            ->recommendation($this->createRecommendationCommand('user-a'))
+            ->recommendation($this->createRecommendationCommand('user-a', null))
```
Comment (OndraM) [inline tests/integration/RequestBuilder/RecommendationRequestBuilderTest.php:24]: I think this should be kept without the second param - to test the "basic" use-case (ie. called wihout optional params).

**Review by OndraM — CHANGES_REQUESTED** [review]

File: src/Model/Command/Sorting.php (hunk 1 above)
Comment (foglcz) [reply src/Model/Command/Sorting.php:20]: On most clients, it will be - until they decide to use different model than default. `model_name` is optional parameter in the payload, allowing us to switch default models server-side.

For example, if the client decides to test default model against new model (`alpha`), and then decides to use the `alpha` as new default, then we won't *rename* the model -- we'd just switch configuration server side to use the new one for all requests that don't explicitly request a different model name.

Changing parameters of existing model needs full recalculation of the model - which is a costly operation. By the time the client decides to change the default model, we'll already have everything calculated & up to date in the new model; so doing a recalc feels kinda pointless if we can just change the default pointer server-side.

File: tests/integration/RequestBuilder/EventsRequestBuilderTest.php (hunk 2 above)
Comment (foglcz) [reply tests/integration/RequestBuilder/EventsRequestBuilderTest.php:18]: ahaha, that's because fresh instances don't have these item properties set up.
I can cherry-pick it into separate PR if you want. I don't really feel it's a problem though.

File: tests/integration/RequestBuilder/EventsRequestBuilderTest.php (hunk 3 above)
Comment (foglcz) [reply tests/integration/RequestBuilder/EventsRequestBuilderTest.php:26]: :+1: will fix, thanks

**Comment by foglcz** [comment]
Closing based on review - this PR will be split into two.

File: src/Model/Command/UserRecommendation.php (hunk 8 above)
Comment (foglcz) [reply src/Model/Command/UserRecommendation.php:57]: agreed :+1:
